use assumptions for integration variables

smichr opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->

fixes #7334

#### Brief description of what is fixed or changed

definite integration limits are used to assign assumptions for the integration variables when they differ from the integration variable.

#### Other comments

The more strict assumption (e.g. positive when limits are nonnegative) is used since whether 0 is contained at a limit or not will not matter. (see @asmeurer comment on linked issue) Since `positive` and not `extended_positive` is used, this will also imply that the variable is finite and, again, the result of the integral should not depend on the inclusion of +/-oo.

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
- integrals
    - assumptions based upon definite integration limits are automatically applied to integration variables
<!-- END RELEASE NOTES -->

smichr commented:
@oscargus , I had made the previous PR online and couldn't update remotely. Here is the same PR with the wester test modified to remove assumptions. Still, reversing the limits makes a big difference on the time it takes to do the integral...that is a different issue, I think.

smichr committed:
limits set assumptions of integration variable

oscargus commented:
I do think it solves #16014 as well. Just trying the added code manually it seems like it does. The issue was really that the assumptions should not have to be done manually and since you could remove them in the test, it means that it solves it. Time for order is another thing, which indeed is a problem as such, but not really what the issue was about.

(I've been thinking about a fix for this recently, but I do not think that the code would have been as simple and clean as your...)

smichr committed:
watch for separate=True result

smichr committed:
update tests

The test for 15716 is replaced with a MWE that would fail
unless the ValueError is caught, but it does not require
integration.

The integration result is otherwise modified since the
real value of x used during integration allows the
log to be split so 3 instead of 2 terms are produced.

smichr committed:
without sorting, check for real

oscargus commented on sympy/integrals/integrals.py:
```diff
@@ -439,6 +439,34 @@ def doit(self, **hints):
+                d = Dummy(positive=True)
+            elif all(i.is_nonpositive for i in l) and not x.is_nonpositive:
+                d = Dummy(negative=True)
+            elif all(i.is_real for i in l) and not x.is_real:
```
Is it a problem having extendedreal here? I guess for integrals with oo bounds it would help.

smichr replied:
I was going off of @asmeurer's comment that the value of an integral at the limit will not matter...so I am just keeping it simple. If you can find a case where it matters we could make the change.

asmeurer replied:
real is more restricted than extended real, so it should lead to more simplifications. Unless we can get a wrong answer from having real here instead of extended real I would just stick with real.

jksuom replied:
> integrals with oo bounds

Such (improper) integrals are limits of integrals taken over finite intervals. The integration variable should not take infinite values.

oscargus replied:
I was primarily considering the type of integral as in the in this PR edited Wester test. Now, it wasn't the y-variable that caused the major part of the simplification, but I guess that there may be cases where this would be the case.

But as you say, easy to change if there are cases where it actually matters.

oscargus replied:
You are off course correct @jksoum, still:
```
$ git grep Integral | grep oo | wc -l
145
```

oscargus replied:
Forgot this:
```
$ git grep integrate | grep oo | wc -l
330
```
